python/dgl/utils/data.py

```diff
@@ -130,46 +149,45 @@ def graphdata2tensors(data, idtype=None, bipartite=False, **kwargs):
     num_dst : int
         Number of destination nodes.
     """
-    if idtype is None and not (isinstance(data, tuple) and F.is_tensor(data[0])):
+    # Convert tuple to SparseAdjTuple
+    if isinstance(data, tuple):
+        if not isinstance(data[0], str):
+            # (row, col) format, convert to ('coo', (row, col))
+            data = ('coo', data)
+        data = SparseAdjTuple(*data)
+
+    if idtype is None and \
+            not (isinstance(data, SparseAdjTuple) and F.is_tensor(data.arrays[0])):
         # preferred default idtype is int64
         # if data is tensor and idtype is None, infer the idtype from tensor
         idtype = F.int64
     checks.check_valid_idtype(idtype)
 
-    if isinstance(data, tuple) and (not F.is_tensor(data[0]) or not F.is_tensor(data[1])):
+    if isinstance(data, SparseAdjTuple) and (not all(F.is_tensor(a) for a in data.arrays)):
         # (Iterable, Iterable) type data, convert it to (Tensor, Tensor)
-        if len(data[0]) == 0:
+        if len(data.arrays[0]) == 0:
             # force idtype for empty list
-            data = F.tensor(data[0], idtype), F.tensor(data[1], idtype)
+            data = SparseAdjTuple(data.format, tuple(F.tensor(a, idtype) for a in data.arrays))
```

Comment:
Did you handle empty edge ID list? Maybe using a for loop is cleaner:
```python
for i in range(len(data.arrays)):
    a = data.arrays[i]
    if F.is_tensor(a):
        continue
    elif len(a) == 0:
        # force idtype for empty list
        data.arrays[i] = F.tensor(a, idtype)
    else:
        # convert the iterable to tensor and keep its native data type so we can check
        # its validity later
        data.arrays[i] = F.tensor(a)
```

Reply:
I don't think you can replace the contents within a tuple.

Reply:
Also it already handles empty list (unit test added).